Mathlib/Tactic/ByContra.lean

```diff
@@ -41,17 +41,14 @@ example : 1 < 2 := by
   -- h : ¬ 1 < 2 ⊢ False
 ```
 -/
-syntax (name := byContra!) "by_contra!" (ppSpace colGt binderIdent)? Term.optType : tactic
+syntax (name := byContra!) "by_contra!" (ppSpace colGt rcasesPatMed)? (" : " term)? : tactic
 
 macro_rules
-  | `(tactic| by_contra! $[: $ty]?) =>
-    `(tactic| by_contra! $(mkIdent `this):ident $[: $ty]?)
-  | `(tactic| by_contra! _%$under $[: $ty]?) =>
-    `(tactic| by_contra! $(mkIdentFrom under `this):ident $[: $ty]?)
-  | `(tactic| by_contra! $e:ident) => `(tactic| (by_contra $e:ident; try push_neg at $e:ident))
-  | `(tactic| by_contra! $e:ident : $y) => `(tactic|
-       (by_contra! h
-        -- if the below `exact` call fails then this tactic should fail with the message
-        -- tactic failed: <goal type> and <type of h> are not definitionally equal
-        have $e:ident : $y := by { (try push_neg); exact h }
-        clear h))
+| `(tactic| by_contra! $[$pat?]? $[: $ty?]?) => do
+  let pat ← pat?.getDM `(rcasesPatMed| $(mkIdent `this):ident)
+  let replaceTac ← match ty? with
+    | some ty => `(tactic| replace h : $ty := by (try push_neg); exact h) -- Let `h` have type `ty`.
+    | none => `(tactic| skip)
+  -- We have to use `revert h; rintro $pat` instead of `obtain $pat := h`,
+  -- because if `$pat` is a variable, `obtain $pat := h` doesn't do anything.
+  `(tactic| (by_contra h; (try push_neg at h); $replaceTac; revert h; rintro ($pat:rcasesPatMed)))
```

Comment:
Thanks for adding a comment, this slowly makes sense now. I wonder, though: why does this do `try push_neg at h` twice? I would think `replaceTac` doesn't need to do this, right?